reconcile.go

```diff
@@ -1,256 +1,216 @@
+}
+func reconcileLogError(log logr.Logger, err error) (*ctrl.Result, error) {
+	return reconcileLogErrorWithResult(log, nil, err)
+}
+func reconcileLogErrorWithResult(log logr.Logger, result *ctrl.Result, err error) (*ctrl.Result, error) {
+	log.Error(err, "Reconcile failed")
+	return result, err
 }
 
-func isDeleting(resource *unstructured.Unstructured) bool {
-	deleting := false
-	metadata, gotMetadata := resource.Object["metadata"].(map[string]interface{})
-	if gotMetadata {
-		if _, deleting = metadata["deletionTimestamp"]; deleting {
-			return true
-		}
-	}
-	return false
+func (r *TerraformReconciler) isDeleting(resource *unstructured.Unstructured) bool {
+	return resource.GetDeletionTimestamp() != nil
 }
-func ensureFinalizer(resource *unstructured.Unstructured) {
-	metadata, gotMetadata := resource.Object["metadata"].(map[string]interface{})
-	if !gotMetadata {
-		metadata = map[string]interface{}{}
-	}
-	finalizers, gotFinalizers := metadata["finalizers"].([]string)
-	if !gotFinalizers {
-		finalizers = []string{}
-	}
+func (r *TerraformReconciler) ensureFinalizer(ctx context.Context, log logr.Logger, resource *unstructured.Unstructured) error {
+	copyResource := resource.DeepCopy()
+
+	finalizers := resource.GetFinalizers()
 	for _, f := range finalizers {
 		if f == "tfoperatorbridge.local" {
 			// already present
-			return
+			log.Info("Finalizer - already exists")
+			return nil
 		}
 	}
+	log.Info("Finalizer - adding")
 	finalizers = append(finalizers, "tfoperatorbridge.local")
-	metadata["finalizers"] = finalizers
-	resource.Object["metadata"] = metadata
-}
-func removeFinalizer(resource *unstructured.Unstructured) {
-	metadata, gotMetadata := resource.Object["metadata"].(map[string]interface{})
-	if !gotMetadata {
-		metadata = map[string]interface{}{}
-	}
-	finalizers, gotFinalizers := metadata["finalizers"].([]string)
-	if !gotFinalizers {
-		finalizers = []string{}
+	resource.SetFinalizers(finalizers)
+
+	if err := r.client.Patch(ctx, resource, client.MergeFrom(copyResource)); err != nil {
+		return fmt.Errorf("Error saving finalizer: %s", err)
 	}
+	return nil
+}
+func (r *TerraformReconciler) removeFinalizerAndSave(ctx context.Context, log logr.Logger, resource *unstructured.Unstructured) error {
+	copyResource := resource.DeepCopy()
+
+	finalizers := resource.GetFinalizers()
+	foundFinalizer := false
 	updatedFinalizers := []string{}
 	for _, f := range finalizers {
 		if f == "tfoperatorbridge.local" {
-			log.Println("Removing finalizer")
+			log.Info("Finalizer - removing")
+			foundFinalizer = true
 		} else {
 			updatedFinalizers = append(updatedFinalizers, f)
 		}
 	}
-	metadata["finalizers"] = updatedFinalizers
-	resource.Object["metadata"] = metadata
+	if !foundFinalizer {
+		log.Info("Finalizer - not found on remove")
+		return fmt.Errorf("Finalizer not found on remove")
+	}
+	resource.SetFinalizers(updatedFinalizers)
+	if err := r.client.Patch(ctx, resource, client.MergeFrom(copyResource)); err != nil {
```

Comment:
I really like this patch approach using the deepcopy

Reply:
Yeah, it feels more targetted. Apart from `finalizers`, I think all the updates are on `status` now which should reduce the chance of conflicts now that it's a subresource.